services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java

```diff
@@ -5894,46 +5894,62 @@ public void killOfflinePush(String clusterName, String kafkaTopic, boolean isFor
-         *
-         * The target use case is for new fabric build out.
-         * In Blueshift project, we are planning to leverage participant store to short-circuit
-         * the bootstrap if the corresponding version has been killed in the source fabric.
-         * So in the new fabric, we need the kill-job message to be as accurate as possible to
-         * avoid the discrepancy.
+         * Check whether the specified store is already terminated or not.
+         * If yes, the kill job message will be skipped.
          */
-        LOGGER.info("Resource: {} has finished bootstrapping, so kill job will be skipped", kafkaTopic);
-        return;
+        Optional<Version> version;
+        try {
+          version = getStoreVersion(clusterName, kafkaTopic);
+        } catch (VeniceNoStoreException e) {
+          LOGGER.warn(
+              "Kill job will be skipped since the corresponding store for topic: {} doesn't exist in cluster: {}",
+              kafkaTopic,
+              clusterName);
+          return;
+        }
+        if (version.isPresent() && VersionStatus.isBootstrapCompleted(version.get().getStatus())) {
+          /**
+           * This is trying to avoid kill job entry in participant store if the version is already online.
+           * This won't solve all the edge cases since the following race condition could still happen, but it is fine.
+           * The race condition could be:
+           * 1. The version status is still not 'ONLINE' yet from Controller's view, then kill job message will be sent to storage node;
+           * 2. When storage node receives the kill job message, the version from storage node's perspective could be already
+           *    'online';
+           *
+           * The reason not to fix this race condition:
+           * 1. It should happen rarely;
+           * 2. This will change the lifecycle of store version greatly, which might introduce other potential issues;
+           *
+           * The target use case is for new fabric build out.
+           * In Blueshift project, we are planning to leverage participant store to short-circuit
+           * the bootstrap if the corresponding version has been killed in the source fabric.
+           * So in the new fabric, we need the kill-job message to be as accurate as possible to
+           * avoid the discrepancy.
+           */
+          LOGGER.info("Resource: {} has finished bootstrapping, so kill job will be skipped", kafkaTopic);
+          return;
+        }
+
+        if (version.isPresent() && VersionStatus.isVersionKilled(version.get().getStatus())) {
+          LOGGER.info("Resource: {} has been updated to KILLED previously, so kill job will be skipped", kafkaTopic);
+          return;
+        }
+
+        // Update version status to KILLED on ZkNode.
+        ReadWriteStoreRepository repository = resources.getStoreMetadataRepository();
```

Comment:
Just wondering if this can go wrong if version.isPresent() is false here?

Maybe we should have a big if clause checking version.isPresent() outside and wrap all logics here.

Reply:
Good point, Jialin! Let's say version.isPresent is false, then it means that the store has not yet had a version created for that version number (kafka topic). In this case, we should avoid the update here, as the version is unavailable. Will fix it.